### PR TITLE
Variables in Custom JS Funtion Variable

### LINKED ISSUE #212: Support variables in Custom JavaScript variable
# Summary

Using an existing variable into a Custom JavaScript variable can avoid a lot of dev and can be really useful.
Example: you wish return a custom attribute from current clicked element. Use `{{ClickElement}}` is essential.

## Current behavior

Into a custom javascript variable, add the `{{ClickElement}}` variable like :

```javascript
function () { if ({{ClickElement}}) { return {{ClickElement}}.getAttribute('my-attribute'); } else { return ""; } };
```

This code return the following error:

Screen Shot 2020-01-06 at 16 09 16
Screen Shot 2020-01-06 at 16 09 33

Moreother, when this kind of error occured, the Preview mode break totally 😕 

Maybe a similar issue here: https://forum.matomo.org/t/how-to-refer-user-defined-variables-in-custom-javascript/34932

## Expected behavior

Add a variable into a custom javascript variable

### PR DESCRIPTION
**Pull request by ulcuber**
### Description:

Resolves #212

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by AltamashShaikh** [comment]
@ulcuber Does this code works for you ?
```
function () { if ({{ClickElement}}) { return {{ClickElement}}.getAttribute('class'); } else { return "111"; } };
```

**Comment by ulcuber** [comment]
@AltamashShaikh checked this issue. It was just creating variable. I forgot to evaluate variable. Now it works

File: Template/Variable/CustomJsFunctionVariable.php [hunk 1]
```diff
@@ -36,30 +36,42 @@ public function getParameters()
                 $field->title = 'JavaScript Function';
                 $field->description = 'The value should start with "function() { " and end with "return yourValue; }". You have to define a function and return a value. We highly recommend to test the pasted JavaScript function to avoid JavaScript errors on your website.';
                 $field->uiControl = FieldConfig::UI_CONTROL_TEXTAREA;
+                $field->customFieldComponent = self::FIELD_TEXTAREA_VARIABLE_COMPONENT;
+                $field->prepare = function ($value) {
```
Comment (AltamashShaikh) [inline Template/Variable/CustomJsFunctionVariable.php:40]: @ulcuber any reason why have we added this line ?

**Review by AltamashShaikh — COMMENTED** [review]
@ulcuber Changes looks good
@tsteur Can you also please review it once ?

File: Template/Variable/CustomJsFunctionVariable.php (hunk 1 above)
Comment (ulcuber) [reply Template/Variable/CustomJsFunctionVariable.php:40]: Added because value was trimmed twice

Comment (AltamashShaikh) [reply Template/Variable/CustomJsFunctionVariable.php:40]: I don't think `prepare` works in anyway, you can simply added the trim back may be ?

**Comment by tsteur** [comment]
@AltamashShaikh @ulcuber the reason I didn't implement it I believe was because I couldn't figure out how to make this work without using `eval`. I believe in some cases this wouldn't work maybe like below?

```js
var {{ foo }} = {{ bar }};
```

Not sure if such cases would work? Or if the variables would only work in few specific places?

File: Template/Variable/CustomJsFunctionVariable.php (hunk 1 above)
Comment (ulcuber) [reply Template/Variable/CustomJsFunctionVariable.php:40]: In which case should it not work?

Comment (AltamashShaikh) [reply Template/Variable/CustomJsFunctionVariable.php:40]: @ulcuber Unable to find any implementation of `prepare` in `https://github.com/matomo-org/matomo/blob/4.x-dev/core/Settings/FieldConfig.php`

Comment (ulcuber) [reply Template/Variable/CustomJsFunctionVariable.php:40]: https://github.com/matomo-org/matomo/blob/48c458c830fd75ab8cc90489d5a31110ee6c36c9/core/Settings/FieldConfig.php#L213
https://github.com/matomo-org/matomo/blob/48c458c830fd75ab8cc90489d5a31110ee6c36c9/core/Settings/Setting.php#L224

Comment (AltamashShaikh) [reply Template/Variable/CustomJsFunctionVariable.php:40]: Hey @ulcuber Thanks for pointing it out, my IDE was not recognising the `prepare` changes dunno why.
I can see now :+1:

**Comment by snake14** [comment]
Closing as I believe that this was implemented by #731 . Thank you for your contribution @ulcuber . Sorry that we didn't get it merged.

**Pull request closed without merging**